Core/MIPS/ARM64/Arm64CompLoadStore.cpp

```diff
@@ -0,0 +1,385 @@
+// addresses in function prologues/epilogues:
+//  sw s5, 104(sp)
+//  sw s4, 100(sp)
+//  sw s3, 96(sp)
+//  sw s2, 92(sp)
+//  sw s1, 88(sp)
+//  sw s0, 84(sp)
+//  sw ra, 108(sp)
+//  mov s4, a0
+//  mov s3, a1
+//  ...
+// Such sequences could easily be detected and turned into nice contiguous
+// sequences of ARM stores instead of the current 3 instructions per sw/lw.
+//
+// Also, if we kept track of the likely register content of a cached register,
+// (pointer or data), we could avoid many BIC instructions.
+
+
+#include "Core/MemMap.h"
+#include "Core/Config.h"
+#include "Core/MIPS/MIPS.h"
+#include "Core/MIPS/MIPSAnalyst.h"
+#include "Core/MIPS/MIPSCodeUtils.h"
+#include "Core/MIPS/ARM64/Arm64Jit.h"
+#include "Core/MIPS/ARM64/Arm64RegCache.h"
+
+#define _RS MIPS_GET_RS(op)
+#define _RT MIPS_GET_RT(op)
+#define _RD MIPS_GET_RD(op)
+#define _FS MIPS_GET_FS(op)
+#define _FT MIPS_GET_FT(op)
+#define _FD MIPS_GET_FD(op)
+#define _SA MIPS_GET_SA(op)
+#define _POS  ((op>> 6) & 0x1F)
+#define _SIZE ((op>>11) & 0x1F)
+#define _IMM16 (signed short)(op & 0xFFFF)
+#define _IMM26 (op & 0x03FFFFFF)
+
+// All functions should have CONDITIONAL_DISABLE, so we can narrow things down to a file quickly.
+// Currently known non working ones should have DISABLE.
+
+// #define CONDITIONAL_DISABLE { Comp_Generic(op); return; }
+#define CONDITIONAL_DISABLE ;
+#define DISABLE { Comp_Generic(op); return; }
+
+namespace MIPSComp
+{
+	using namespace Arm64Gen;
+	using namespace Arm64JitConstants;
+	
+	// Destroys SCRATCH2
+	void Arm64Jit::SetScratch1ToEffectiveAddress(MIPSGPReg rs, s16 offset) {
+		if (offset) {
+			ADDI2R(SCRATCH1, gpr.R(rs), offset, SCRATCH2);
+		} else {
+			MOV(SCRATCH1, gpr.R(rs));
+		}
+	}
+
+	void Arm64Jit::SetCCAndSCRATCH1ForSafeAddress(MIPSGPReg rs, s16 offset, ARM64Reg tempReg, bool reverse) {
+		SetScratch1ToEffectiveAddress(rs, offset);
+		// TODO
+	}
+
+	void Arm64Jit::Comp_ITypeMemLR(MIPSOpcode op, bool load) {
+		CONDITIONAL_DISABLE;
+		int offset = (signed short)(op & 0xFFFF);
+		MIPSGPReg rt = _RT;
+		MIPSGPReg rs = _RS;
+		int o = op >> 26;
+
+		DISABLE;
+		// TODO: For some reason I can't get this to work on ARM64.
+		if (!js.inDelaySlot && false) {
+			// Optimisation: Combine to single unaligned load/store
+			bool isLeft = (o == 34 || o == 42);
+			MIPSOpcode nextOp = Memory::Read_Instruction(js.compilerPC + 4);
+			// Find a matching shift in opposite direction with opposite offset.
+			if (nextOp == (isLeft ? (op.encoding + (4 << 26) - 3)
+				: (op.encoding - (4 << 26) + 3))) {
+				EatInstruction(nextOp);
+				nextOp = MIPSOpcode(((load ? 35 : 43) << 26) | ((isLeft ? nextOp : op) & 0x03FFFFFF)); //lw, sw
+				Comp_ITypeMem(nextOp);
+				return;
+			}
+		}
+
+		u32 iaddr = gpr.IsImm(rs) ? offset + gpr.GetImm(rs) : 0xFFFFFFFF;
+		bool doCheck = false;
+		FixupBranch skip;
+
+		if (gpr.IsImm(rs) && Memory::IsValidAddress(iaddr)) {
+			u32 addr = iaddr & 0x3FFFFFFF;
```

Comment:
I guess this masking should probably be using a constant, hmm.  Anyway, it's not necessary here in 64-bit right?  I wonder if it's better with or without for SetRegImm...

-[Unknown]